### PR TITLE
feat: mod the hash before checking it is a valid point

### PR DESCRIPTION
**Pull request by girazoki**
This PR modifies the hashtotry algorithm. The modulus in the curve BN256 G1 is quite low, and values above the modulus are considered invalid points. In consequence the hashtotry performs quite a few iterations before finding a valid point. We decided to perform a H(message||ctr) mod modulus before checking whether it yields a valid point, increasing the odds of finding valid points in less iterations.

### CODE REVIEW

File: src/bn256/mod.rs [hunk 1]
```diff
@@ -69,15 +71,32 @@ impl Bn256 {
...
-            // Check validity of `H` (i.e. point exists in group G1)
-            self.arbitrary_string_to_g1(&attempted_hash).ok()
+            let hash = &self.calculate_sha256(&v)[0..32];
+            // this should never fail as the length of sha256 is max 256
+            let mut attempted_hash = arith::U256::from_slice(hash).unwrap();
+
+            // the library does not provide a function to do a modulus
+            // we use the provided add function adding a 0
+            // we also need to iterate here as the library does the modulus only once
+            while attempted_hash > Fq::modulus() {
```
Comment (tmpolaczyk) [inline src/bn256/mod.rs]: Here we should check for modulo bias, if `attempted_hash >= last_multiple_of_fq_modulus_lower_than_2_256` we should retry the hash with the next counter value.

Comment (girazoki) [reply src/bn256/mod.rs]: fixing it

**Comment by aesedepece** [comment]
From what I read, this is a breaking change, and should trigger a new release, right?

File: src/bn256/mod.rs [hunk 2]
```diff
@@ -48,6 +48,8 @@ impl Bn256 {
     fn arbitrary_string_to_g1(&self, data: &[u8]) -> Result<G1, Error> {
         let mut v = vec![0x02];
         v.extend(data);
+        println!("intofun{:x?}", v);
```
Comment (aesedepece) [inline src/bn256/mod.rs]: _Side effected functions_ :smirk:

**Review by aesedepece — CHANGES_REQUESTED** [review]

File: src/bn256/mod.rs [hunk 3]
```diff
@@ -69,15 +76,35 @@ impl Bn256 {
...
+            // this should never fail as the length of sha256 is max 256
+            let mut attempted_hash = arith::U256::from_slice(hash).unwrap();
+
+            if attempted_hash > LAST_MULTIPLE_OF_FQ_MODULUS_LOWER_THAN_2_256 {
+                return None;
+            }
+            // the library does not provide a function to do a modulus
+            // we use the provided add function adding a 0
+            // we also need to iterate here as the library does the modulus only once
+            while attempted_hash > Fq::modulus() {
```
Comment (tmpolaczyk) [inline src/bn256/mod.rs]: Can you extract this while loop into a separate function? Something like `u256_to_modulus`

File: src/bn256/mod.rs [hunk 4]
```diff
@@ -69,15 +76,35 @@ impl Bn256 {
...
+                attempted_hash.add(&arith::U256::zero(), &Fq::modulus());
+            }
+            let mut s = [0u8; 32];
+            match attempted_hash.to_big_endian(&mut s) {
+                Ok(_) => self.arbitrary_string_to_g1(&s).ok(),
+                _ => None,
+            }
```
Comment (aesedepece) [inline src/bn256/mod.rs]: Would this be possible?
```suggestion
            attempted_hash.to_big_endian(&mut s).and_then(|_| self.arbitrary_string_to_g1(&s))
```

**Review by aesedepece — CHANGES_REQUESTED** [review]

File: src/bn256/mod.rs [hunk 5]
```diff
@@ -69,15 +76,35 @@ impl Bn256 {
...
         let point = c.find_map(|ctr| {
             v[position] = ctr;
-            let attempted_hash = self.calculate_sha256(&v);
-            // Check validity of `H` (i.e. point exists in group G1)
-            self.arbitrary_string_to_g1(&attempted_hash).ok()
+            let hash = &self.calculate_sha256(&v)[0..32];
+            // this should never fail as the length of sha256 is max 256
+            let mut attempted_hash = arith::U256::from_slice(hash).unwrap();
+
+            if attempted_hash > LAST_MULTIPLE_OF_FQ_MODULUS_LOWER_THAN_2_256 {
```
Comment (tmpolaczyk) [inline src/bn256/mod.rs]: :eyes: 

```suggestion
            if attempted_hash >= LAST_MULTIPLE_OF_FQ_MODULUS_LOWER_THAN_2_256 {
```

Comment (tmpolaczyk) [inline src/bn256/mod.rs]: Add comment about modulo bias

**Review by aesedepece — APPROVED** [review]

File: src/bn256/mod.rs [hunk 6]
```diff
@@ -36,6 +42,29 @@ use error::Error;
...
+    ///
+    /// # Arguments
+    ///
+    /// * `num` - the number we want to reduce.
+    /// * `modulus` - the modulus we want to apply.
+    ///
+    /// # Returns
+    ///
+    /// * If successful, a `U256` representing num % modulus.
+
```
Comment (tmpolaczyk) [inline src/bn256/mod.rs]: Remove this empty line. And since this function does not need self, you can move it outside of `impl Bn256` and remove `&self`.

**Review by tmpolaczyk — APPROVED** [review]

**Comment by tmpolaczyk** [comment]
Remember that we need to publish a new version to crates.io in order to see the new algorithm in the node.

**Pull request merged**